flash/transport_qspi_samd.go

```diff
@@ -0,0 +1,179 @@
+	}
+	return ErrInvalidClockSpeed
+}
+
+func (q qspi) runCommand(cmd Command) (err error) {
+	QSPI.INSTRCTRL.Set(uint32(cmd))
+	QSPI.INSTRFRAME.Set(QSPI_INSTRFRAME_WIDTH_SINGLE_BIT_SPI |
+		QSPI_INSTRFRAME_ADDRLEN_24BITS | QSPI_INSTRFRAME_INSTREN |
+		(QSPI_INSTRFRAME_TFRTYPE_READ << QSPI_INSTRFRAME_TFRTYPE_Pos))
+	QSPI.INSTRFRAME.Get() // dummy read for synchronization, as per datasheet
+	q.endTransfer()
+	return
+}
+
+func (q qspi) readCommand(cmd Command, buf []byte) (err error) {
+	q.disableAndClearCache()
+	QSPI.INSTRCTRL.Set(uint32(cmd))
+	const iframe = QSPI_INSTRFRAME_WIDTH_SINGLE_BIT_SPI | QSPI_INSTRFRAME_DATAEN |
+		QSPI_INSTRFRAME_ADDRLEN_24BITS | QSPI_INSTRFRAME_INSTREN |
+		(QSPI_INSTRFRAME_TFRTYPE_READ << QSPI_INSTRFRAME_TFRTYPE_Pos)
+	QSPI.INSTRFRAME.Set(iframe)
+	QSPI.INSTRFRAME.Get() // dummy read for synchronization, as per datasheet
+	var ptr uintptr = qspi_AHB_LO
+	for i := range buf {
+		buf[i] = volatile.LoadUint8((*uint8)(unsafe.Pointer(ptr)))
+		ptr++
+	}
+	q.endTransfer()
+	q.enableCache()
+	return
+}
+
+func (q qspi) readMemory(addr uint32, buf []byte) (err error) {
+	if (addr + uint32(len(buf))) > (qspi_AHB_HI - qspi_AHB_LO) {
+		return ErrInvalidAddrRange
+	}
+	q.disableAndClearCache()
+	QSPI.INSTRCTRL.Set(uint32(CmdQuadRead))
+	const iframe = QSPI_INSTRFRAME_WIDTH_QUAD_OUTPUT | QSPI_INSTRFRAME_DATAEN |
+		QSPI_INSTRFRAME_ADDRLEN_24BITS | QSPI_INSTRFRAME_INSTREN |
+		QSPI_INSTRFRAME_ADDREN | (8 << QSPI_INSTRFRAME_DUMMYLEN_Pos) |
+		(QSPI_INSTRFRAME_TFRTYPE_READMEMORY << QSPI_INSTRFRAME_TFRTYPE_Pos)
+	QSPI.INSTRFRAME.Set(iframe)
+	QSPI.INSTRFRAME.Get() // dummy read for synchronization, as per datasheet
+	ln := len(buf)
+	sl := (*[1 << 28]byte)(unsafe.Pointer(uintptr(qspi_AHB_LO + addr)))[:ln:ln]
+	copy(buf, sl)
+	q.endTransfer()
+	q.enableCache()
+	return
+}
+
+func (q qspi) writeCommand(cmd Command, data []byte) (err error) {
+	iframe := uint32(QSPI_INSTRFRAME_WIDTH_SINGLE_BIT_SPI |
+		QSPI_INSTRFRAME_ADDRLEN_24BITS | QSPI_INSTRFRAME_INSTREN |
+		(QSPI_INSTRFRAME_TFRTYPE_WRITE << QSPI_INSTRFRAME_TFRTYPE_Pos))
+	if len(data) > 0 {
+		iframe |= QSPI_INSTRFRAME_DATAEN
+	}
+	q.disableAndClearCache()
+	QSPI.INSTRCTRL.Set(uint32(cmd))
+	QSPI.INSTRFRAME.Set(iframe)
+	QSPI.INSTRFRAME.Get() // dummy read for synchronization, as per datasheet
+	var ptr uintptr = qspi_AHB_LO
+	for i := range data {
+		volatile.StoreUint8((*uint8)(unsafe.Pointer(ptr)), data[i])
+		ptr++
+	}
+	q.endTransfer()
+	q.enableCache()
+	return
+}
+
+func (q qspi) eraseCommand(cmd Command, address uint32) (err error) {
+	panic("implement me")
+}
+
+func (q qspi) writeMemory(addr uint32, data []byte) (err error) {
+	panic("implement me")
+}
+
+//go:inline
+func (q qspi) enableCache() {
+	CMCC.CTRL.SetBits(CMCC_CTRL_CEN)
+}
+
+//go:inline
+func (q qspi) disableAndClearCache() {
+	CMCC.CTRL.ClearBits(CMCC_CTRL_CEN)
+	for CMCC.SR.HasBits(CMCC_SR_CSTS) {
+	}
+	CMCC.MAINT0.SetBits(CMCC_MAINT0_INVALL)
+}
+
+//go:inline
```

Comment:
`//go:inline` is something that should usually be left to the compiler. The compiler will do that almost always for very short functions and functions that are only called from one place.

Did you see a performance problem without `//go:inline`?

Reply:
I did see some measured improvement in another driver, but I did not measure anything for this use case.  I was just cargo-culting what I had done previously, so I will remove this.

Reply:
removed